clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1061,16 +1082,36 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
                 requestRejoin();
                 future.raise(error);
             } else if (error == Errors.ILLEGAL_GENERATION) {
-                log.info("Attempt to heartbeat failed since generation {} is not current", sentGeneration.generationId);
-                resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
-                future.raise(error);
+                if (generationUnchanged()) {
+                    log.info("Attempt to heartbeat failed since current {} is not valid, resetting generation", sentGeneration);
+                    resetGenerationOnResponseError(ApiKeys.HEARTBEAT, error);
+                    future.raise(error);
+                } else {
+                    // if the generation has changed, then ignore this error
+                    log.info("Attempt to heartbeat failed since old {} is not valid, ignoring the error", sentGeneration);
+                    future.complete(null);
+                }
             } else if (error == Errors.FENCED_INSTANCE_ID) {
-                log.error("Received fatal exception: group.instance.id gets fenced");
-                future.raise(error);
+                if (generationUnchanged()) {
+                    log.info("Attempt to heartbeat failed since current {} gets fenced with group instance id {}",
+                        sentGeneration, rebalanceConfig.groupInstanceId);
+                    future.raise(error);
+                } else {
+                    // if the generation has changed, then ignore this error
```

Comment:
Not sure why we ignore fenced errors here but not in the JoinGroup/SyncGroup handlers.

Reply:
As explained in the PR description, I intentionally make the logic differently: the rationale is that for join / sync we should only have one request in-flight at a given time, and at that time if the generation has changed it should be from the heartbeat error handling that resets it; for this case this client should have not re-joined and replaced its member.id, so if the error returns it still indicates another member with the same instance.id has replaced the member.id.

For heartbeat though, we've observed it is possible that, 1) heartbeat sent, 2) the same member rejoins group, gets a new member.id, 3) the previous heartbeat is handled and the error set. In this case we should not treat it as fatal.